cpp/ql/src/experimental/Security/CWE/CWE-297/ValidationCertificateHostMismatch.ql

```diff
@@ -0,0 +1,73 @@
+/**
+ * @name Validation certificate host mismatch.
+ * @description Lack of certificate name validation allows any valid certificate to be used. And that can lead to disclosure.
+ * @kind problem
+ * @id cpp/validation-certificate-host-mismatch
+ * @problem.severity warning
+ * @precision medium
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-297
+ */
+
+import cpp
+import semmle.code.cpp.valuenumbering.GlobalValueNumbering
+
+predicate checkHostnameOlder1() {
+  exists(FunctionCall fctmp1 |
+    fctmp1.getTarget().hasName("X509_get_ext_d2i") and
+    (
+      exists(MacroInvocation mi1tmp, MacroInvocation mi2tmp |
+        mi1tmp.getMacroName() = "sk_GENERAL_NAME_num" and
+        mi2tmp.getMacroName() = "sk_GENERAL_NAME_value"
+      )
+      or
+      exists(FunctionCall fctmp2, FunctionCall fctmp3 |
+        fctmp2.getTarget().hasName("sk_GENERAL_NAME_num") and
+        fctmp3.getTarget().hasName("sk_GENERAL_NAME_value")
+      )
+    )
+  )
+}
+
+predicate checkHostnameOlder2() {
```

Comment:
Same question here: This predicate is a property of the entire codebase that either holds or doesn't hold. And I can make it hold by adding a couple of calls to some functions anywhere in the codebase. Please describe what are you trying to do here, and I'm sure we can fix it up.

Reply:
I have implemented a search for all possible ways to validate the name of the certificate specified in https://wiki.openssl.org/index.php/Hostname_validation.

I'm looking for code where the developer doesn't use any of these name validation methods at all, but uses certificate validation. In this case, it's definitely a threat.
I skip when there is an error in the implementation of one of the methods, or the method is not implemented for all connections.